# BEASTIE
BEASTIE (Bayesian Estimation of Allele Specific Transcription Integrating across Exons) is a software suite for identifying allele-specific-expression (ASE) from regulatory variants from RNA-seq and WGS data.
BEASTIE uses a Bayesian hierarchical model to integrate prior information with read count data and genetic data. Using MCMC (Markov Chain Monte Carlo), BEASTIE efficiently performs posterior inference to estimate effect sizes of ASE. <br>
BEASTIE has been found to be substantially more accurate than other tests based on the binomial distribution.

## Description
<place holder>
<figure>
  
## Download
BEASTIE can be downloaded from XXXX
  
## Installation
### Prerequisites for installing BEASTIE
The following are required to install and run BIRD directly on your system:
* BEASTIE has been tested on **Linux**. It may or may not work on other UNIX systems.
* [CmdStan](https://mc-stan.org/users/interfaces/cmdstan) must be installed.  This is the command-line interface to the STAN statistical programming language.
* [Python](https://www.python.org/downloads/release/python-360/) version 3.6 or higher is required.
  
### Installing and Compiling BEASTIE source code
<place holder>
  
## Running BEASTIE
### Prerequisites
<place holder>
  
### Input Format
<place holder>

## Running the model
### Output Format
<place holder>

## Example Inputs and Outputs
<place holder>

  
  
